### PR TITLE
Self timed readback output node

### PR DESCRIPTION
**Pull request by ervanalb**

### CODE REVIEW

File: src/EffectNode.cpp [hunk 1]
```diff
@@ -15,105 +15,91 @@
...
 
-EffectNode::EffectNode(Context *c, QString file)
-    : VideoNode(c)
-    , m_intensity(0)
-    , m_openGLWorker(new EffectNodeOpenGLWorker(this), &QObject::deleteLater)
-    , m_ready(false) {
+EffectNode::EffectNode(Context *context, QString file)
+    : VideoNode(new EffectNodePrivate(context))
+{
+    qRegisterMetaType<Chain>("Chain");
```
Comment (zbanks) [inline src/EffectNode.cpp]: Why is this here?

File: src/EffectNode.cpp [hunk 2]
```diff
@@ -15,105 +15,91 @@
...
-    o.insert("intensity", m_intensity);
+    o.insert("file", d()->m_file);
+    o.insert("intensity", d()->m_intensity);
     return o;
 }
 
-void EffectNode::onInitialized() {
+void EffectNode::chainsEdited(QList<Chain> added, QList<Chain> removed) {
+    for (auto chain : added) {
+        auto result = QMetaObject::invokeMethod(d()->m_openGLWorker.data(), "addNewState", Q_ARG(Chain, chain));
```
Comment (zbanks) [inline src/EffectNode.cpp]: I don't see why we have to go through the MetaObject here: isn't `m_openGLWorker.data()` known to be a `EffectNodeOpenGLWorker`? Can't you just call `m_openGLWorker->addNewState(chain)`?

File: src/EffectNode.cpp [hunk 3]
```diff
@@ -15,105 +15,91 @@
...
+    qreal intensityIntegral;
+    qreal frequency;
+    {
+        QMutexLocker locker(&d()->m_stateLock);
+        if (!d()->m_ready) {
+            //qDebug() << this << "is not ready";
+            return inputTextures.at(0); // Pass-through
+        }
+        //qDebug() << "Looking up" << chain << "in" << d()->m_renderStates << "of" << this;
+        if (!d()->m_renderStates.contains(chain)) { // This check doesn't seem to work
```
Comment (zbanks) [inline src/EffectNode.cpp]: What does this comment mean, it sounds bad

File: src/EffectNode.cpp (hunk 2 above)
Comment (zbanks) [reply src/EffectNode.cpp]: oh I see, it's to do message passing across threads. I almost wish that could be more explicit but this is fine.

File: src/SelfTimedReadBackOutputNode.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MAG_FILTER,GL_LINEAR);
+    glTexParameteri(GL_TEXTURE_2D,GL_TEXTURE_MIN_FILTER,GL_LINEAR);
+
+    m_shader->setUniformValue("iFrame", 0);
+
+    vao->bind();
+    glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+    vao->release();
+
+    glReadPixels(0, 0, m_size.width(), m_size.height(), GL_RGBA, GL_UNSIGNED_BYTE, m_pixelBuffer.data());
```
Comment (zbanks) [inline src/SelfTimedReadBackOutputNode.cpp]: And this is fine because it's from an FBO, not a texture?

File: src/ImageNode.cpp [hunk 5]
```diff
@@ -10,166 +10,172 @@
...
-            for(auto i = 0; i < m_openGLWorker->m_frameDelays.size(); ++i) {
-                if(m_openGLWorker->m_frameDelays.at(i) >= extraMs)
-                    return m_openGLWorker->m_frameTextures.at(i)->textureId();
+    auto extraMs = currentMs;
+    if (totalDelay) {
+        extraMs = currentMs % totalDelay;
+        if (frameDelays.empty()) {
+            for (auto i = 0; i < frameDelays.size(); ++i) {
+                if (frameDelays.at(i) >= extraMs)
+                    return frameTextures.at(i)->textureId();
```
[the remark this answers is not in the document]
Comment (ervanalb) [reply src/ImageNode.cpp]: I didn't actually change behavior here, I've actually never tried the existing code on anything besides nyancat...

File: src/Model.cpp [hunk 6]
```diff
@@ -24,39 +29,49 @@ bool Edge::operator==(const Edge &other) const {
...
-    return m_chains;
+QList<Chain> Model::chains() {
+    return d_ptr->m_chains;
 }
 
 void Model::prepareNode(VideoNode *videoNode) {
     if(!videoNode)
         return;
 
+    // Giving a QSP a parent... hmm...
```
[the remark this answers is not in the document]
Comment (ervanalb) [reply src/Model.cpp]: It's weird because they are two different ways Qt has for managing object lifetimes, and you don't really mix them. Nothing technically wrong with this, just feels bad

File: src/VideoNode.h [hunk 7]
```diff
@@ -132,10 +115,23 @@ protected slots:
...
+};
+
+class VideoNodePrivate : public QObject {
+    Q_OBJECT
+
+public:
+    VideoNodePrivate(Context *context);
+
+    int m_inputCount{};
+    QMutex m_stateLock; // TODO this is no longer a meaningful concept, should use separate locks for separate fields
```
[the remark this answers is not in the document]
Comment (ervanalb) [reply src/VideoNode.h]: yeah, we can take separate locks for separate bits of data for gainzz

File: src/EffectNode.cpp (hunk 1 above)
Comment (ervanalb) [reply src/EffectNode.cpp]: this is here so it can be passed via `invokeMethod` as an argument (Q_ARG)

File: src/EffectNode.cpp (hunk 2 above)
Comment (ervanalb) [reply src/EffectNode.cpp]: yeah, I only use `invokeMethod` to traverse threads. Polymorphism should work fine for calling things in the same thread.

File: src/EffectNode.cpp (hunk 3 above)
Comment (ervanalb) [reply src/EffectNode.cpp]: This comment existed in the original code, I didn't change it. I can delete it?

File: src/SelfTimedReadBackOutputNode.cpp (hunk 4 above)
Comment (ervanalb) [reply src/SelfTimedReadBackOutputNode.cpp]: Yeah, `ReadPixels` is much nicer about the returned data not having padding and stuff

File: src/ImageNode.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ervanalb) [reply src/ImageNode.cpp]: Yeah, confirmed that this is fixed

File: src/EffectNode.cpp (hunk 3 above)
Comment (ervanalb) [reply src/EffectNode.cpp]: Confirmed, this check does in fact work
